File: emessgee/subscriber.py
import atexit

from emessgee.memory_queue import ReadMemoryQueue
from emessgee.exceptions import TopicDoesNotExistError, ErrorMessages

class Subscriber:
    def __init__(self, topic:str):
        self._topic = topic
        self._topic_queues = {
            topic: ReadMemoryQueue(topic)
        }
        self._queue_index = 0
         
        atexit.register(self.close)
    
    def recv(self, topic):
        if(topic not in self._topic_queues.keys()):
            raise TopicDoesNotExistError(
                ErrorMessages.TOPIC_DOES_NOT_EXIST.format(topics=self._topic_queues.keys())
            ) 
        
        return self._topic_queues[topic].read()

    def close(self):
        for topic_queue in self._topic_queues.values():
            topic_queue.close()

File: src/subscriber.cpp
#include <subscriber.h>

namespace emessgee
{

Subscriber::Subscriber(std::string topic)
{
    _topic_queues.insert({
        topic,
        std::make_unique<ReadMemoryQueue>(topic)
    });
}

Subscriber::Subscriber(std::vector<std::string> topics)
{

    for(std::string& topic : topics)
    {
        _topic_queues.insert({
            topic,
            std::make_unique<ReadMemoryQueue>(topic)
        });
    }
}

Subscriber::~Subscriber()
{
    close();
}

ReadResult Subscriber::recv(std::string topic)
{
    if(_topic_queues.count(topic) > 0)
    {
        return _topic_queues[topic]->read();
    }

    return ReadResult();
}

void Subscriber::close()
{
    std::map<std::string, ReadMemoryQueue::Ptr>::iterator it;

    for(it = _topic_queues.begin(); it != _topic_queues.end(); ++it)
    {
        it->second->close();
    }
}

}